### DISCUSSION TITLE
Pre-Assessment - Building a Monolithic Library Management System API

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/spring-framework#145] [Tasks]
**Task: Library Management System API Development**

You have been assigned to develop a RESTful API for a Library Management System using Spring Boot. The API will allow users to perform various operations related to managing books, authors, genres, library branches, and borrower information. Follow the steps below to complete the task:

1. Project Setup:
   - Create a new Spring Boot project using your preferred IDE or Spring Initializr.
   - Configure the project with the necessary dependencies, such as Spring Web, Spring Data JPA, and MySQL Connector.
   - Set up a MySQL database and configure the connection properties in the `application.properties` file.

2. Entity Design:
   - Design the entity classes for the following entities: `Book`, `Author`, `Genre`, `LibraryBranch`, and `Borrower`.
   - Each entity should have appropriate attributes and relationships based on the requirements.

3. Data Persistence:
   - Implement the necessary JPA annotations to map the entity classes to database tables.
   - Define the relationships between entities, such as many-to-many between `Book` and `Author` and many-to-many between `Book` and `Genre`.
   - Create the repository interfaces by extending the appropriate Spring Data JPA interfaces.

4. API Endpoints:
   - Implement the API endpoints to handle the following operations:
     - Book-related endpoints:
       - Retrieve all books: `GET /api/books`
       - Get a book by ID: `GET /api/books/{id}`
       - Add a new book: `POST /api/books`
       - Update a book: `PUT /api/books/{id}`
       - Delete a book: `DELETE /api/books/{id}`
     - Author-related endpoints:
       - Retrieve all authors: `GET /api/authors`
       - Get an author by ID: `GET /api/authors/{id}`
       - Add a new author: `POST /api/authors`
       - Update an author: `PUT /api/authors/{id}`
       - Delete an author: `DELETE /api/authors/{id}`
     - Genre-related endpoints:
       - Retrieve all genres: `GET /api/genres`
       - Get a genre by ID: `GET /api/genres/{id}`
       - Add a new genre: `POST /api/genres`
       - Update a genre: `PUT /api/genres/{id}`
       - Delete a genre: `DELETE /api/genres/{id}`
     - Library branch-related endpoints:
       - Retrieve all library branches: `GET /api/library-branches`
       - Get a library branch by ID: `GET /api/library-branches/{id}`
       - Add a new library branch: `POST /api/library-branches`
       - Update a library branch: `PUT /api/library-branches/{id}`
       - Delete a library branch: `DELETE /api/library-branches/{id}`
     - Borrower-related endpoints:
       - Retrieve all borrowers: `GET /api/borrowers`
       - Get a borrower by ID: `GET /api/borrowers/{id}`
       - Add a new borrower: `POST /api/borrowers`
       - Update a borrower: `PUT /api/borrowers/{id}`
       - Delete a borrower: `DELETE /api/borrowers/{id}`
     - Additional endpoints:
       - Retrieve books by genre: `GET /api/books/genre/{id}`
       - Retrieve books by author: `GET /api/books/author/{id}`
       - Retrieve books by library branch: `GET /api/books/library-branch/{id}`
       - Retrieve books borrowed by a borrower: `GET /api/books/borrower/{id}`

5. Error Handling:
   - Implement appropriate error handling mechanisms for different scenarios, such as handling invalid input, non-existent resources, and database errors.
   - Return meaningful error messages and appropriate HTTP status codes.

6. Unit Testing  (Optional):
   - Write unit tests for the service layer and controller layer using JUnit and Mockito.
   - Test different scenarios to ensure the correct behavior of the API endpoints.

7. Documentation  (Optional):
   - Generate API documentation using Swagger or Spring REST Docs.
   - Document the API endpoints, request/response structures, and example usage.

8. Additional Functionality (Optional):
   - Implement pagination and sorting for retrieving lists of entities.
   - Add search functionality to allow users to search for books by title, authors, genres, etc.
   - Implement additional business rules or features based on your creativity.

9. Deployment  (Optional):
   - Deploy the application to a cloud platform like Heroku or AWS Elastic Beanstalk.

**Implementation Guidelines:**

- Use Java and Spring Boot to implement the RESTful API.
- Follow the principles of SOLID design and clean code practices.
- Use the appropriate design patterns and architectural principles.
- Utilize Spring Data JPA for data persistence and Spring MVC for handling HTTP requests and responses.
- Apply validation and error handling mechanisms to ensure the robustness of the API.
- Use appropriate status codes and error messages in the API responses.
- Write unit tests for the service layer and controller layer to ensure the correctness of the implementation.
- Document the API endpoints, request/response structures, and example usage.
- Provide clear instructions for setting up the project, running the application, and accessing the API endpoints.
- Use Git for version control and commit your code regularly.

Remember to provide clear documentation, including instructions for setting up the project, running the application, and any additional considerations or assumptions made during development.

**Comment by kuttiram** [comment]
[librarymanagement.zip](https://github.com/akash-coded/spring-framework/files/11594983/librarymanagement.zip)

Hi Akash,

This is Ramesh Kumar Devasundaram, Can you please review the attached code and provide you feedback

**Comment by hardik061** [comment]
[spring-boot-library-management-app.zip](https://github.com/akash-coded/spring-framework/files/11595124/spring-boot-library-management-app.zip)

Hello Akash,

Please find attached the application code. Additional endpoints part is still WIP.

Regards
Hardik Agrawal

**Comment by vinayak355** [comment]
[library-management-system-with-lombok-and-logging.zip](https://github.com/akash-coded/spring-framework/files/11602371/library-management-system-with-lombok-and-logging.zip)

Monolithic approach - with logging and lombok
Please check.
Next, I will segregate each microservice

**Comment by kuttiram** [comment]
[LibraryManagement_MicorService.zip](https://github.com/akash-coded/spring-framework/files/11602785/LibraryManagement_MicorService.zip)

Added another solution zip file.
1. Implemented Microservice approaches (BookingSerivce, GenreService & AuthorService)
2. Implemented Logging & Swagger Docs
Please check and let me know your feedback

Thanks & Regards,
Ramesh Kumar D

**Comment by Kailashp700** [comment]
Kindly go through my code and let me know if there needs any improvement.
It is not fully completed borrower and additional part is pending
[LibraryManagement.zip](https://github.com/akash-coded/spring-framework/files/11604210/LibraryManagement.zip)

Thanks and Regards,
Kailash Prajapati

**Reply by Kailashp700** [reply]
[Library_925to.zip](https://github.com/akash-coded/spring-framework/files/11620397/Library_925to.zip)
Kindly go through my code I have made some changes let let me know where i can make improvement

**Comment by jay4tech** [comment]
@akash-coded 
Name - Jay Prakash Kumar

Here is the Pre-Assessment Part 2  - LibraryManagement Monolithic

https://github.com/jay4tech/microservices-architecture/tree/pre-assessment-part-2-library-management-monolithic

Swagger UI -
http://localhost:8080/swagger-ui/index.html

**Comment by arghyagiri** [comment]
@akash-coded
Solution link -
https://github.com/arghyagiri/microservice-e2/blob/main/library-management-system/Read.md

**Reply by shriramram** [reply]
@akash-coded 
Name - Dattatraya Shriram
Description - Lib management with only Book api and others is WIP.(Partial implementation)
[lib-management-app.zip](https://github.com/akash-coded/spring-framework/files/12755226/lib-management-app.zip)

**Comment by Rajashree-Ravi** [comment]
@akash-coded 
Name - Rajashree R
Please find below the link for pre-assessment submission

https://github.com/Rajashree-Ravi/library-management-system/tree/main/src/main/java/com/example/library/service/implementation

**Comment by Sourav21071998** [comment]
Please find the attachment of monolithic library management.
Name: Sourav Sarkar

[MonolithicLibraryManagement.zip](https://github.com/akash-coded/spring-framework/files/12782971/MonolithicLibraryManagement.zip)

**Comment by akratimishra22** [comment]
[library-management-system.zip](https://github.com/akash-coded/spring-framework/files/12785261/library-management-system.zip)

**Reply by akratimishra22** [reply]
By Akrati Mishra

**Reply by akratimishra22** [reply]
https://drive.google.com/file/d/1rrBgtqC58-pVmtZrf_EGHW_fb4u3K9e9/view?usp=drive_web

Above is the link for ecommerce assignment, Im not able to find the git discussion for it, also not able to upload the folder as it is more than 25mb. Please consider this as submission for Thursday's ecommerce assignment.

**Comment by samkrish16** [comment]
Hi Aakash,
Please find my pre-assessment of Monolithic Library Management
Name:Sampath Kumar
Employee Id:759553
Org:TCS

https://drive.google.com/file/d/1QGaxO7EHXOu7Okjpy0P-HmwXjOIkqC4o/view?usp=sharing

**Comment by sayanp-12** [comment]
Hi @akash-coded , Please find below GitHub link for Pre-assessment library management solution

Name : Sayan Panja

https://github.com/sayanp-12/edureca_library_management

**Reply by sayanp-12** [reply]
hi @akash-coded , please find the solution of Order management  project

[spring-boot-ecommerce-joins_jpa_sayan.zip](https://github.com/akash-coded/spring-framework/files/12795047/spring-boot-ecommerce-joins_jpa_sayan.zip)

**Comment by chrismjac** [comment]
Hi @akash-coded , Please find below GitHub link for Pre-assessment library management solution

Name : Chris Mary Jacob

https://github.com/chrismjac/libraryManagement

**Comment by Manogaran-Coded** [comment]
Hi @akash-coded - Please find the Screenshot document for Library Managment Project
[Library Management Service.docx](https://github.com/akash-coded/spring-framework/files/13226794/Library.Management.Service.docx)

**Comment by HarshDPandya** [comment]
Hi @akash-coded , please find attached zip file for Monolithic Library Management System API.
Name : Harsh Pandya
[library-management-HarshPandya-2066481.zip](https://github.com/akash-coded/spring-framework/files/13277209/library-management-HarshPandya-2066481.zip)

**Comment by jayakumarvinayagam2023** [comment]
Hi Akash,
I have completed the Book end point only using .Net Core, C#, SQL Server, I will continue the rest.

Book-related endpoints:
    Retrieve all books: GET /api/books
    Get a book by ID: GET /api/books/{id}
    Add a new book: POST /api/books
    Update a book: PUT /api/books/{id}
    Delete a book: DELETE /api/books/{id}
**_https://github.com/jayakumarvinayagam2023/-MonolithicLibraryManagementSystem/tree/main_**

**Comment by samik2023** [comment]
Hi Aakash,

Here is the preasseement code attached.

[Samik- TCS - Nov 7th Batch]
[demolib.zip](https://github.com/akash-coded/spring-framework/files/13291318/demolib.zip)


Thanks
Samik

**Comment by bijumedayil** [comment]
[LibraryManager.zip](https://github.com/akash-coded/spring-framework/files/13291649/LibraryManager.zip)

No Jar attached because of Size
Swagger url 

http://localhost:8080/api/swagger-ui/index.html
Bijumon Habeeb
TCS

**Comment by maheshgajera** [comment]
Hi Akash,

Please find the source code for library management
[library-management-MaheshGajera-1952963.zip](https://github.com/akash-coded/spring-framework/files/13291670/library-management-MaheshGajera-1952963.zip)


Regards,
Mahesh Gajera

**Comment by sathyakulandhaivel** [comment]
Hi Akash,
Please find my pre-assessment of Monolithic Library Management
[Monolithic Library Mgmt_SathyaK_563920.zip](https://github.com/akash-coded/spring-framework/files/13293287/Monolithic.Library.Mgmt_SathyaK_563920.zip)

**Comment by himarou** [comment]
Hi Akash,
Please find my pre-assessment of Monolithic Library Management.

Thanks,
Himansu
[Himansu-LMS-project.zip](https://github.com/akash-coded/spring-framework/files/13299626/Himansu-LMS-project.zip)

**Comment by sowmyagardas** [comment]
Hi Akash,
I have attached my assignment on Library management system Monolithic application. Please check and let me know if anything needs to be improved.

As I'm new to this, have taken only author and book as entities. Executed GET/POST/PUT/DELETE using Postman and it worked.


Thanks,
Sowmya.
[library-management-system-main.code-sowmya_workspace.code-workspace.zip](https://github.com/akash-coded/spring-framework/files/13347580/library-management-system-main.code-sowmya_workspace.code-workspace.zip)

**Comment by jyothi241868** [comment]
Hi Akash,

Attaching my assignment on Library Management system. Please verify and let me know if any suggestions.

Thanks,
Jyothi K

[LibraryManagementSystem-Monolythic.zip](https://github.com/akash-coded/spring-framework/files/13540485/LibraryManagementSystem-Monolythic.zip)

**Comment by raghavan-tcs** [comment]
@akash-coded  Hi Aakash,

 Please find attached pre-assessment for Library management system. for 18-April-2024 Microservices Developer Batch

[librarymanagementsystem.zip](https://github.com/akash-coded/spring-framework/files/15038390/librarymanagementsystem.zip)


Thanks
Raghavan S

**Comment by SumitShokeen41** [comment]
Hi Akash,

Please find pre-assessment on Library Management system in Attachments. Microservices Developer Batch (17-April-2024)

Thanks,
Sumit Shokeen
[LibraryManagement.zip](https://github.com/akash-coded/spring-framework/files/15038957/LibraryManagement.zip)

**Comment by Aditi-12092024** [comment]
Hi Akash,
Please find below my pre-assessment on Library management System. Microservices Developer Batch (19-April-2024)


[LibraryManagement.zip](https://github.com/akash-coded/spring-framework/files/15039095/LibraryManagement.zip)
Thanks

**Comment by nari93** [comment]
Hi Akash,

Please find the solution file on pre-assessment on library management system for batch: 17 April 2024.
[LibraryManagement.zip](https://github.com/akash-coded/spring-framework/files/15040187/LibraryManagement.zip)



Best regards,
Naresh P.

**Comment by hyma24** [comment]
Hi @akash-coded 

Please find the attached pre-assessment on library management system for the batch (17-04-2023).
[librarymanagement.zip](https://github.com/akash-coded/spring-framework/files/15040180/librarymanagement.zip)

Regards,
Hymavathi

**Comment by umangpathak123** [comment]
Hello @akash-coded 

PFA Pre-Assessment. 

Thanks
Umang 
[lms.zip](https://github.com/akash-coded/spring-framework/files/15041756/lms.zip)

**Comment by venom-xd** [comment]
Hello @akash-coded 

Please find the link of pre-assessment on library management system for the batch (17-04-2023).

Thanks
Aditya Kumar

https://drive.google.com/file/d/1ihwnBaPY5-MYD4-QpwLxywkj8n3Vpk1s/view?usp=sharing

**Comment by OnewayShruti** [comment]
Hello @akash-coded  , 
library management system for the batch (17-04-2023)

[Pre-assesment.zip](https://github.com/akash-coded/spring-framework/files/15080410/Pre-assesment.zip)

shruti vanave - 1956049

**Comment by prasetha21-max** [comment]
***PRASETHA N***
[Pre-Assesement.zip](https://github.com/user-attachments/files/20868551/Pre-Assesement.zip)

**Comment by OnewayShruti** [comment]
Hello
Could you let me know the details for this assignment.

**Reply by akash-coded** [reply]
Sure. You can bring it up in the connect. I'll elaborate.

**Comment by OnewayShruti** [comment]
Hi Akash ,

Ok, thanks. Could you let me know the details about the connect I would be
interested to pass it on to my friends

Regards
Shruti Vanave
